File: eggrally_v1.py
import pyxel
import json
import math
import random
from dataclasses import dataclass
from typing import Callable

#phase 0

@dataclass
class Rectangle:
    x: float
    y: float
    width: int
    height: int
    color: int

    @property
    def top(self): return self.y
    @property
    def bottom(self): return self.y + self.height
    @property
    def left(self): return self.x
    @property
    def right(self): return self.x + self.width
    @property
    def midx(self): return self.x + self.width // 2
    @property
    def midy(self): return self.y + self.height // 2

@dataclass
class Egg(Rectangle):
    current_hp: int
    max_hp: int

    # @property
    # def current_hp(self): return self.current_hp

    def move(self, speed: int):
        if pyxel.btn(pyxel.KEY_W):
            self.y -= speed
        if pyxel.btn(pyxel.KEY_S):
            self.y += speed
        if pyxel.btn(pyxel.KEY_A):
            self.x -= speed
        if pyxel.btn(pyxel.KEY_D):
            self.x += speed
        self.x = max(0, min(self.x, pyxel.width - self.width))
        self.y = max(0, min(self.y, pyxel.height - self.height))

@dataclass
class Eggnemy(Rectangle):
    def move_towards(self, egg: Egg, speed:int):
        target_x = egg.x + egg.midx
        target_y = egg.y + egg.midy

        center_x = self.x + self.midx
        center_y = self.y + self.midy

        diff_x = target_x - center_x
        diff_y = target_y - center_y

        distance = math.sqrt( diff_x**2 + diff_y**2)

        self.x += diff_x / distance * speed
        self.y += diff_y / distance * speed

@dataclass
class Config:
    screen_width: int
    screen_height: int 
    fps: int
    eggnemy_count: int
    egg_speed: int = 5
    eggnemy_speed: int = 2
    egg_color: int = 7 #white
    eggnemy_color: int = 8 # red

# MODELL
class Model:
    def __init__(self, config: Config, egg: Egg, eggnemies: list[Eggnemy]):
        self._config: Config = config
        self._egg: Egg = egg
        self._eggnemies: list[Eggnemy] = eggnemies
        self._is_game_over: bool = False
        self._score: int = 0
        self._ticks: int = 0
        self._last_hit: int = -1000

    def update(self):
        if self._is_game_over:
            return

        self._ticks += 1
        egg_speed = self._config.egg_speed
        eggnemy_speed = self._config.eggnemy_speed
        self._egg.move(egg_speed)

        for enem in self._eggnemies:
            enem.move_towards(self._egg, eggnemy_speed)

            if self._is_colliding(self._egg, enem):
                if self._ticks - self._last_hit >= self._config.fps: 
                    self._egg.current_hp -= 1
                    self._last_hit = self._ticks

                    if self._egg.current_hp <= 0:
                        self._is_game_over = True
                        break
        
        if pyxel.btn(pyxel.KEY_L):
            self._attack()

    def _is_colliding(self, rect1: Rectangle, rect2: Rectangle):
        return not (
            rect1.right < rect2.left or
            rect1.left > rect2.right or
            rect1.bottom < rect2.top or
            rect1.top > rect2.bottom
        )

    def _attack(self):
        attack_range = 50  # hard coded
        target_x = self._egg.midx
        target_y = self._egg.midy

        remaining_enemies: list[Eggnemy] = []
        for enem in self._eggnemies:
            enem_center_x = enem.midx
            enem_center_y = enem.midy

            distance = math.sqrt((enem_center_x - target_x) ** 2 + (enem_center_y - target_y) ** 2)
            
            if distance > attack_range:
                remaining_enemies.append(enem)
            else:
                self._score += 1

        self._eggnemies = remaining_enemies

    @property
    def egg_width(self):
        return self._egg.width
 
    @property
    def egg_height(self):
        return self._egg.height
 
    @property
    def fps(self):
        return self._config.fps

    @property
    def egg(self) -> Egg:
        return self._egg
    
    @property
    def eggnemies(self) -> list[Eggnemy]:
        return self._eggnemies
    
    @property
    def score(self) -> int:
        return self._score
 
    @property
    def is_game_over(self) -> bool:
        return self._is_game_over
        
#VIEW
class View:
    def __init__(self, width: int, height: int):
        self._screen_width: int = width
        self._screen_height: int = height

    def start(self, update: Callable[[], None], draw: Callable[[], None], fps:int):
        pyxel.init(self._screen_width, self._screen_height, fps=fps)
        pyxel.run(update, draw)
    
    def clear_screen(self):
        pyxel.cls(0)

    def draw_egg(self, egg: Egg):
        pyxel.rect(egg.x, egg.y, egg.width, egg.height, egg.color) 
    
    def draw_eggnemy(self, eggnemy: Eggnemy):
        pyxel.rect(eggnemy.x, eggnemy.y, eggnemy.width, eggnemy.height, eggnemy.color)  #red

    def draw_egg_current_hp(self, egg: Egg):
        pyxel.text(egg.midx - 5, egg.bottom + 5, f"{str(egg.current_hp)}/{str(egg.max_hp)}", egg.color)

    def draw_score(self, score: int):
        pyxel.text(self._screen_width // 2 - 3, 10, f"Score{str(score)}", 7) #??

class Controller:
    def __init__(self, model: Model, view: View):
        self._model = model
        self._view = view
    
    def start(self):
        self._view.start(self.update, self.draw, self._model.fps)

    def update(self):
        self._model.update()
    
    def draw(self):
        self._view.clear_screen()
        self._view.draw_score(self._model.score)
        if not self._model.is_game_over:
            self._view.draw_egg(self._model.egg)
            self._view.draw_egg_current_hp(self._model.egg)
        for enemy in self._model.eggnemies:
            self._view.draw_eggnemy(enemy)

if __name__ == "__main__":
    with open("settings.json", "r") as f:
        settings = json.load(f)

    config = Config(
        screen_width = settings["worldWidth"],
        screen_height = settings["worldHeight"],
        fps = settings["fps"],
        eggnemy_count = settings["numberOfEggnemies"]
    )
    egg: Egg = Egg(
        x = config.screen_width // 2,
        y = config.screen_height // 2,
        width = settings["eggWidth"],
        height = settings["eggHeight"],
        current_hp = settings["eggInitialHP"],
        max_hp = settings["eggInitialHP"],
        color = config.egg_color
    )
    eggnemies: list[Eggnemy] = []
    for i in range(config.eggnemy_count):
        eggnemies.append(Eggnemy(
            x = random.randint(0, config.screen_width - settings["eggnemyWidth"]), # random spawn
            y = random.randint(0, config.screen_height - settings["eggnemyHeight"]),
            width = settings["eggnemyWidth"],
            height = settings["eggnemyHeight"],
            color = config.eggnemy_color
        ))

    model = Model(config, egg, eggnemies)
    view = View(config.screen_width, config.screen_height)
    controller = Controller(model, view)
    controller.start()

File: eggrally_v1.ts
import { Cmd, h, startModelCmd } from "cs12242-mvu/src"
import { CanvasMsg, canvasView } from "cs12242-mvu/src/canvas"
import * as Canvas from "cs12242-mvu/src/canvas"
import { Match as M, Schema as S, pipe } from "effect"

// phase 0

const getSettings = async () => {
    const result = await fetch("settings.json")
    //if (!result.ok) throw new Error("Failed to load settings.json") //guards if .json file dont be
    return result.json()
}

const spawnEggnemies = (settings): Eggnemy[] =>
    Array.from({ length: settings.numberOfEggnemies }, () => {
        const maxX = settings.worldWidth - settings.eggnemyWidth
        const maxY = settings.worldHeight - settings.eggnemyHeight

        return {
            x: Math.floor(Math.random() * maxX),
            y: Math.floor(Math.random() * maxY),
            width: settings.eggnemyWidth,
            height: settings.eggnemyHeight,
        }
    })

const attack = (model: Model): Eggnemy[] => 
    model.eggnemies.filter((enem) => 
        !isColliding(model.egg, enem))
    



const Egg = S.Struct({
    x: S.Number,
    y: S.Number,
    width: S.Number,
    height: S.Number,
    currentHP: S.Number,
    maxHP: S.Number,
})
type Egg = typeof Egg.Type

const Eggnemy = S.Struct({ 
    x: S.Number,
    y: S.Number,
    width: S.Number,
    height: S.Number,
})
type Eggnemy = typeof Eggnemy.Type

const Config = S.Struct({
    screenWidth: S.Int,
    screenHeight: S.Int,
    fps: S.Int,
    canvasID: S.String,
})
type Config = typeof Config.Type

const Model = S.Struct({
    config: Config,
    egg: Egg,
    eggnemies: S.Array(Eggnemy),
    isGameOver: S.Boolean,
    score: S.Int,
    ticks: S.Int,
    lastHit: S.Int,
})
type Model = typeof Model.Type

// UPDATE
type Msg = CanvasMsg

const Update = (msg: Msg, model: Model): Model => 
    M.value(msg).pipe(
        M.tag("Canvas.MsgKeyDown", ({ key }) => {
            if (model.isGameOver) return model

            const [x, y] = move(model, key)

            const newEggnemies = key === "l" ? attack(model) : model.eggnemies

            return {
                ...model,
                egg: {
                    ...model.egg,
                    x,
                    y,
                },
                eggnemies: newEggnemies,
            }
        }),
        M.tag("Canvas.MsgTick", () => {
            const gettingHit = model.eggnemies.some(enem => isColliding(model.egg, enem))
            let newHP =  model.egg.currentHP
            let lastHit = model.lastHit
            if (!model.isGameOver && gettingHit && model.ticks + 1 - model.lastHit >= model.config.fps ) {
                newHP = model.egg.currentHP - 1
                lastHit = model.ticks + 1
            }
            const gameOver = newHP <= 0

            return {
                ...model,
                egg: {
                    ...model.egg,
                    currentHP: newHP,
                },
                eggnemies: gameOver ? model.eggnemies : moveEggnemies(model),
                isGameOver: gameOver,
                ticks: model.ticks + 1,
                lastHit: lastHit,
        }}),
        M.orElse(() => (model))
    )

const move = (model: Model, key: String) => {
    const moveSpeed = 10
    const maxX = model.config.screenWidth - model.egg.width
    const maxY = model.config.screenHeight - model.egg.height

    const x =
        key === "a" || key === "ArrowLeft"
        ? model.egg.x - moveSpeed < 0
            ? model.egg.x
            : model.egg.x - moveSpeed
        : key === "d" || key === "ArrowRight"
            ? model.egg.x + moveSpeed > maxX
            ? model.egg.x
            : model.egg.x + moveSpeed
            : model.egg.x

    const y = 
        key === "w" || key === "ArrowUp"
        ? model.egg.y - moveSpeed < 0
            ? model.egg.y
            : model.egg.y - moveSpeed
        : key === "s" || key === "ArrowDown"
            ? model.egg.y + moveSpeed > maxY
            ? model.egg.y
            : model.egg.y + moveSpeed
            : model.egg.y
    return [x,y]

}

const moveEggnemies = (model: Model): Eggnemy[] => {
    const targetX = model.egg.x + model.egg.width / 2
    const targetY = model.egg.y + model.egg.width / 2

    return model.eggnemies.map(enem => {
        const centerX = enem.x + enem.width /2
        const centerY = enem.y + enem.height / 2

        const diffX = targetX - centerX
        const diffY = targetY - centerY

        return {
            ...enem,
            x: enem.x + diffX / calcD(diffX, diffY),
            y: enem.y + diffY / calcD(diffX, diffY),
        }
    })}

const square = (n: number): number => (n*n)

const calcD = (diffX: number, diffY: number) => {
    const dist = Math.sqrt(square(diffX) + square(diffY))
    return dist === 0 ? 1 : dist
}

const isColliding = (egg: Egg, eggnemy: Eggnemy): boolean => {
    return (
        egg.x < eggnemy.x + eggnemy.width &&
        egg.x + egg.width > eggnemy.x &&
        egg.y < eggnemy.y + eggnemy.height &&
        egg.y + egg.height > eggnemy.y
    )
}

//VIEW
const View = (model: Model) => [
    Canvas.Clear.make({ 
        color: "black"
    }),
    ...(model.egg.currentHP > 0 
        ? [
            Canvas.SolidRectangle.make({
                x: model.egg.x,
                y: model.egg.y,
                width: model.egg.width,
                height: model.egg.height,
                color: "white",
            }),
            Canvas.Text.make({ // egg stats
                x: model.egg.x + model.egg.width / 2 , // center
                y: model.egg.y + model.egg.height + 12,  // below
                text: `${model.egg.currentHP}/${model.egg.maxHP}`,
                fontSize: 12,
                color: "white",
                textAlign: "center",
            }),
        ] : []),
    ...model.eggnemies.map(enem => 
        Canvas.SolidRectangle.make({
            x: enem.x,
            y: enem.y,
            width: enem.width,
            height: enem.height,
            color: "yellow",
    })),
    Canvas.Text.make({
        x: 20,
        y: 20,
        text: "EGGRALLY",
        fontSize: 20,
        color: "white",
        textAlign: "left",
    }),
    Canvas.Text.make({
        x: 20,
        y: 40,
        text: `Score: ${model.score}`,
        fontSize: 16,
        color: "white",
        textAlign: "left",
    }),
]

// START
const root = document.getElementById("app")!

getSettings().then(settings => {
    const initModel: Model = {
        config: {
            screenWidth: settings.worldWidth,
            screenHeight: settings.worldHeight,
            fps: settings.fps,
            canvasID: "canvas",
        },
        egg: {
            x: settings.worldWidth / 2 , // center
            y: settings.worldHeight / 2,
            width: settings.eggWidth,
            height: settings.eggHeight,
            currentHP: settings.eggInitialHP,
            maxHP: settings.eggInitialHP,
        },
        eggnemies: spawnEggnemies(settings),
        isGameOver: false,
        score: 0,
        ticks: 0,
        lastHit: -1000,
    }

    startModelCmd(
        root,
        initModel,
        Update,
        canvasView(
            settings.worldWidth,
            settings.worldHeight,
            settings.fps,
            settings.canvasID,
        View
        )
    )
})
